### PR TITLE
Contraction Hierarchy precomputation

### PR DESCRIPTION
**Pull request by SChudakov**
Precomputation approach is originally described [here](http://algo2.iti.kit.edu/schultes/hwy/contract.pdf).
Parallel version is studied in [this article](https://algo2.iti.kit.edu/download/vetter_sa.pdf)

Performance is measured for road maps of Andorra (|V|=160.478, |E|=325.658) and Liechtenstein (|V|=151.421, |E|=319.712).

Precomputation time.

Andorra_precomputation

Liechtenstein_precomputation

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by jkinable** [comment]
@Toptachamann what's the status of this PR? Would be nice to have it completed. Could you do a first pass. I'd like to have a look at it before a merge.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithmTest.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+import java.util.Random;
+
+import static org.jgrapht.alg.shortestpath.ContractionHierarchyAlgorithm.ContractionEdge;
+import static org.jgrapht.alg.shortestpath.ContractionHierarchyAlgorithm.ContractionVertex;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for the {@link ContractionHierarchyAlgorithm}..
```
Comment (Toptachamann) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithmTest.java]: "Algorithm}.." -> "Algorithm}.". Also, you can state the authorship here

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,1154 @@
...
+
+    /**
+     * Submits {@code tasks} to the {@code completionService} setting start and end
+     * of the working segment and consumer for them
+     *
+     * @param segmentStart start of working segment inclusively
+     * @param segmentEnd   start of working segment exclusively
+     * @param consumer     consumer
+     */
+    private void submitTasks(int segmentStart, int segmentEnd, Consumer<ContractionVertex<V>> consumer) {
```
Comment (Toptachamann) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: There's another method with the same name, but its logic is a bit different. Let's name them differently

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,1154 @@
...
+ * hierarchy. Finally after all vertices are contracted each edge is set to be either upward if its source has
+ * lower level that its target, or downward if vice versa.
+ *
+ * <p>
+ * Computing initial priorities, independent sets and shortcuts, updating neighbours priorities and
+ * marking upward edges is performed in parallel what gives this implementation performance speedup
+ * comparing to the sequential approach.
+ *
+ * <p>
+ * For parallelization, this implementation relies on the {@link ExecutorService}.
```
[the remark this answers is not in the document]
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: This line is analogous to the one in the DeltaSteppingShortestPath. It provides information about how the parallelism is implemented in this algorithm without necessity to read the code.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java [hunk 4]
```diff
@@ -0,0 +1,1154 @@
...
+    }
+
+
+    /**
+     * Computes shortcuts for vertex {@code vertex} wrt the overlay graph.
+     *
+     * @param vertex a vertex in {@code contractionGraph}
+     * @return list of shortcuts
+     */
+    private List<Pair<ContractionEdge<E>, ContractionEdge<E>>> getShortcuts(ContractionVertex<V> vertex) {
```
[the remark this answers is not in the document]
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: This function is used in a lambda. So I would prefer to keep it that way in order for the lambda to remain concise.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java [hunk 5]
```diff
@@ -0,0 +1,1154 @@
...
+     * found. To prune the search keeps track of the value $d(u, v) + max {c(v, w) : (v, w) \in E^{\prime}}$,
+     * where $d(u,v)$ denotes distance between vertex $u$ and $v$, $c(v,w)$ is weight of the edge $(v,w)$,
+     * $E^{\prime}$ is the set of edges of the overlay graph. If the original graph is undirected each
+     * predecessor of {@code vertex} is considered only once and for each found shortcut in the forward
+     * direction another one in the backward direction is generated.
+     *
+     * @param vertex           a vertex in {@code contractionGraph}
+     * @param shortcutConsumer consumer to supply shortcuts to
+     */
+    private void iterateShortcuts(ContractionVertex<V> vertex,
```
[the remark this answers is not in the document]
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: This function iterates shortcuts which corresponds to the vertex and apply the provided shortcutConsumer to every one of them. So I suggest the name of the function together with its arguments are descriptive enough. But I am still open to any suggestions from your side.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java (hunk 2 above)
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: This naming was chosen to emphasize the fact that method overloading technique is used in this case.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Toptachamann) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: Good point

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Toptachamann) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: Can be renamed to 'iterateShortcutEdges'

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithmTest.java [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+import static org.jgrapht.alg.shortestpath.ContractionHierarchyAlgorithm.ContractionEdge;
+import static org.jgrapht.alg.shortestpath.ContractionHierarchyAlgorithm.ContractionVertex;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for the {@link ContractionHierarchyAlgorithm}.
+ */
+public class ContractionHierarchyAlgorithmTest {
```
Comment (Toptachamann) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithmTest.java]: Discussed offline, need to add tests on bigger (~50 vertices) graphs with subsequent basic invariant checking.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Toptachamann) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyAlgorithm.java]: "Parallelization" -> "Parallelisation"

**Comment by jkinable** [comment]
Any progress on this one?

**Comment by SChudakov** [comment]
The latest commit addresses the issues pointed out by @Toptachamann.

**Comment by Toptachamann** [comment]
Thanks for this amazing piece of work!

**Pull request merged**